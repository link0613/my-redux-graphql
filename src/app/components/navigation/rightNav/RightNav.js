// @flow

// #region imports
import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import RightNavButton from './rightNavButton/RightNavButton';
// #endregion

// #region flow types
export type Props = {
  rightLinks: Array<{
    link: string,
    label: string,
    viewName: string,
    btnLink: {
      showWhenUserAuth: boolean,
      alwaysShows: boolean,
    },
  }>,
  onRightNavButtonClick: (event: SyntheticEvent<>) => any,
  userIsAuthenticated: boolean,
};
// #endregion

const RightNav = ({
  rightLinks,
  onRightNavButtonClick,
  userIsAuthenticated,
}: Props) => (
  <ul className="nav navbar-nav navbar-right">
    {userIsAuthenticated
      ? rightLinks
          .filter(btnLink => btnLink.showWhenUserAuth === true)
          .map((aLinkBtn, index) => {
            return (
              <RightNavButton
                key={index}
                link={aLinkBtn.link}
                label={aLinkBtn.label}
                viewName={aLinkBtn.view}
                onClick={onRightNavButtonClick}
              />
            );
          })
      : rightLinks
          .filter(
            btnLink =>
              btnLink.showWhenUserAuth === false ||
              btnLink.alwaysShows === true,
          )
          .map((aLinkBtn, index) => {
            return (
              <RightNavButton
                key={index}
                link={aLinkBtn.link}
                label={aLinkBtn.label}
                viewName={aLinkBtn.view}
                onClick={onRightNavButtonClick}
              />
            );
          })}
  </ul>
);

// #region static props
RightNav.propTypes = {
  rightLinks: PropTypes.arrayOf(
    PropTypes.shape({
      link: PropTypes.string,
      label: PropTypes.string,
      viewName: PropTypes.string,
    }),
  ),
  onRightNavButtonClick: PropTypes.func,
  userIsAuthenticated: PropTypes.bool.isRequired,
};

RightNav.displayName = 'RightNav';
// #endregion

export default RightNav;
