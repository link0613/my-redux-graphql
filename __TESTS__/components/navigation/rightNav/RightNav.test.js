import React from 'react';
import RightNav from '../../../../src/app/components/navigation/rightNav/RightNav';
import renderer from 'react-test-renderer'; // needed both for snpashot testing but also to prevent errors from enzyme
import { MemoryRouter } from 'react-router';

describe('RightNav (navigation) component', () => {
  it('renders as expected (not authenticated)', () => {
    const props = {
      onRightNavButtonClick: () => {},
      userIsAuthenticated: false,
      rightLinks: [
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: false, alwaysShows: true },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: true, alwaysShows: false },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: true, alwaysShows: true },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: false, alwaysShows: false },
        },
      ],
    };
    const component = renderer
      .create(
        <div>
          <MemoryRouter>
            <RightNav {...props} />
          </MemoryRouter>
        </div>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders as expected (authenticated)', () => {
    const props = {
      onRightNavButtonClick: () => {},
      userIsAuthenticated: true,
      rightLinks: [
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: false, alwaysShows: true },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: true, alwaysShows: false },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: true, alwaysShows: true },
        },
        {
          link: 'link',
          label: 'label',
          viewName: '/',
          btnLink: { showWhenUserAuth: false, alwaysShows: false },
        },
      ],
    };
    const component = renderer
      .create(
        <div>
          <MemoryRouter>
            <RightNav {...props} />
          </MemoryRouter>
        </div>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
