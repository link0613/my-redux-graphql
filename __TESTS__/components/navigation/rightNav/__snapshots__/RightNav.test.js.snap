// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightNav (navigation) component renders as expected (authenticated) 1`] = `
<div>
  <ul
    className="nav navbar-nav navbar-right"
  />
</div>
`;

exports[`RightNav (navigation) component renders as expected (not authenticated) 1`] = `
<div>
  <ul
    className="nav navbar-nav navbar-right"
  />
</div>
`;
